samples/glean/src/androidTest/java/org/mozilla/samples/glean/pings/BaselinePingTest.kt

```diff
@@ -0,0 +1,134 @@
+import org.json.JSONObject
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.mozilla.samples.glean.getPingServerAddress
+import java.util.concurrent.TimeUnit
+
+@RunWith(AndroidJUnit4::class)
+class BaselinePingTest {
+    @get:Rule
+    val activityRule: ActivityTestRule<MainActivity> = ActivityTestRule(MainActivity::class.java)
+
+    @get:Rule
+    val gleanRule = GleanTestRule(
+        ApplicationProvider.getApplicationContext(),
+        Configuration(serverEndpoint = getPingServerAddress())
+    )
+
+    @Before
+    fun clearWorkManager() {
+        val context = ApplicationProvider.getApplicationContext<Context>()
+        WorkManagerTestInitHelper.initializeTestWorkManager(context)
+    }
+
+    private fun waitForPingContent(
+        pingName: String,
+        maxAttempts: Int = 3
+    ): JSONObject?
+    {
+        val server = getPingServer()
+
+        var attempts = 0
+        do {
+            attempts += 1
+            val request = server.takeRequest(20L, TimeUnit.SECONDS)
+            val docType = request.path.split("/")[3]
+            if (pingName == docType) {
+                return JSONObject(request.body.readUtf8())
+            }
+        } while (attempts < maxAttempts)
+
+        return null
+    }
+
+    /**
+     * Sadly, the WorkManager still requires us to manually trigger the upload job.
+     * This function goes through all the active jobs by Glean (there should only be
+     * one!) and triggers them.
+     */
+    private fun triggerEnqueuedUpload() {
+        // The tag is really internal to PingUploadWorker, but we can't do much more
+        // than copy-paste unless we want to increase our API surface.
+        val tag = "mozac_service_glean_ping_upload_worker"
+        val reasonablyHighCITimeoutMs = 5000L
+
+        runBlocking {
+            withTimeout(reasonablyHighCITimeoutMs) {
+                do {
+                    val workInfoList = WorkManager.getInstance().getWorkInfosByTag(tag).get()
+                    workInfoList.forEach { workInfo ->
+                        if (workInfo.state === WorkInfo.State.ENQUEUED) {
+                            // Trigger WorkManager using TestDriver
+                            val testDriver = WorkManagerTestInitHelper.getTestDriver()
+                            testDriver.setAllConstraintsMet(workInfo.id)
+                            return@withTimeout
+                        }
+                    }
+                } while (true)
+            }
+        }
+    }
+
+    @Test
+    fun validateBaselinePing() {
+        // Wait for the app to be idle/ready.
+        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
+        val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+        device.waitForIdle()
+
+        // Wait for 1 second: this should guarantee we have some valid duration in the
+        // ping.
+        Thread.sleep(1000)
```

Comment:
:/ (but I get it)

Reply:
Yeah :( The other approach would have been to use internal functions, which I've tried hard to avoid here.